activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb

```diff
@@ -409,7 +409,7 @@ def add_transaction_record(record, ensure_finalize = true)
       end
 
       # Begins the transaction (and turns off auto-committing).
-      def begin_db_transaction()    end
+      def begin_db_transaction(mode = nil)    end
```

Comment:
`begin_db_transaction` is public API I'd rather not change it.

I'd prefer to add a `begin_deferred_db_transaction` that is `# :nodoc:` and only for internal use. On adapters other than `Sqlite3` it would just be an alias to `begin_db_transaction`.

It's also more consistent with `begin_isolated_db_transaction`

Reply:
We need a way to compose `deferred` and `isolated` transactions. You can see this need in the changed test in `activerecord/test/cases/adapters/sqlite3/transaction_test.rb`. Simply having `begin_deferred` and `begin_isolated` doesn't quite provide enough for such composition. I see the point of not wanting to change the public API of `begin_db_transaction`. Maybe an alternative is to move away (first with backwards compat) from `begin_*_db_transaction` methods and add attr_accessors for different transaction modifiers that `begin_db_transaction` uses to craft the desired transaction?

Reply:
> We need a way to compose deferred and isolated transactions. 

If app developers have to concern themselves about whether a transaction should be deffered or immediate, then I'm not sure deferred is a suitable default.

That's not a concern I'm personally willing to expose to users, even more so because it has no equivalent on other adapters. To me that's a deal breaker. If someone else feel strongly about it I'm not gonna oppose, but I won't merge that.

Reply:
App developers will _rarely_ need to concern themselves with whether a transaction should be deferred or immediate, in much the same way they rarely need to concern themselves with whether a transaction should be joinable or not. The main situation in which an app developer would need to concern himself is in fact shared between these concerns: nested transactions. And the test in `activerecord/test/cases/adapters/sqlite3/transaction_test.rb` demonstrates an example:
```ruby
test "opens a `read_uncommitted` transaction" do
  with_connection(flags: shared_cache_flags) do |conn1|
    conn1.create_table(:zines) { |t| t.column(:title, :string) } if in_memory_db?
    conn1.transaction do
      conn1.transaction_manager.materialize_transactions
      conn1.execute("INSERT INTO zines (title) VALUES ('foo')")

      with_connection(flags: shared_cache_flags) do |conn2|
        conn2.transaction(joinable: false, isolation: :read_uncommitted) do
        conn2.transaction(joinable: false, isolation: :read_uncommitted) do
          assert_not_empty(conn2.execute("SELECT * FROM zines WHERE title = 'foo'"))
        end
      end
      raise ActiveRecord::Rollback
    end
  end
end
```

Now, it should be rare indeed that a developer needs a _real_ (joinable: false), _isolated_, _nested_ transaction. But if they ever do, that transaction **must** be _deferred_ when the adapter is SQLite. Otherwise, the database will immediately throw an error and the code will never execute.

I agree to not exposing `mode: :deferred | :immediate` directly through the `#transaction` method. I see your point. My point is only that we cannot simply add a `begin_deferred_db_transaction` method that acts as a sibling of the `begin_isolated_db_transaction` method naively, because sometimes you will need to begin _both_ a `deferred` _and_ an `isolated` transaction (like in the test case above).

So, the naive code that I originally wrote in response to your review isn't sufficient:
```ruby
class RealTransaction < Transaction
  def materialize!
    if isolation_level
      connection.begin_isolated_db_transaction(isolation_level)
    elsif !joinable?
      connection.begin_deferred_db_transaction
    else
      connection.begin_db_transaction
    end

    super
  end
  
  # ...
end
```

This code doesn't allow that test case to pass. We aren't exposing anything about SQLite's transaction modes to the app developer; the test case looks precisely as it does now on `main`, but it does fail and will always fail unless we can set up a transaction that is both deferred and isolated. 

I wanted to share that realization with you as I made it, and share the first thought that come to my mind for how to slightly evolve the codebase to accommodate composing transaction behaviors, instead of simply branching between them. I will finish my update to this branch today and push my suggestion for how this could look. You can then review that code. Sorry for adding confusion by talking about a problem without pushing some code that provides at least one possibility of a solution.

- - -

Small addendum: Postgres (I haven't studied MySQL to know either way) has a similar concept as deferred transactions, it just isn't something that is a concern of ActiveRecord:

> The `DEFERRABLE` transaction property has no effect unless the transaction is also `SERIALIZABLE` and `READ ONLY`. When all three of these properties are selected for a transaction, the transaction may block when first acquiring its snapshot, after which it is able to run without the normal overhead of a `SERIALIZABLE` transaction and without any risk of contributing to or being canceled by a serialization failure. This mode is well suited for long-running reports or backups.
— https://www.postgresql.org/docs/16/sql-set-transaction.html

Not suggesting that we should in fact used `mode:` to the `#transaction` method, but only sharing that this concept isn't completely unique to SQLite. It is only unique to SQLite how _essential_ the difference between the two modes is.

Reply:
>  in much the same way they rarely need to concern themselves with whether a transaction should be joinable or not. 

Again. `joinable` isn't a public property, and public API doesn't allow you to request a non-joinable transaction, it's a purely internal concept. So the comparison doesn't hold.

I'm fine with the adapter having some internal logic for when to use immediate or deferred based on wehther it's nested or other criteria, but the user shouldn't have to think about it.

Reply:
Ok, I think I finally understand what you mean by "public". I was originally interpreting that as "not publicly accessible", but I am seeing that you mean instead "not publicly documented". Gotcha. Again, sorry for my confusion. I think we are on the same page now as to the public interface for `#transaction`, and I will get my code finished today or this weekend so you can review my attempt to support that interface and the internal mechanics it requires.

Thanks again for the helpful reviews. I appreciate it.

Reply:
@byroot: I have introduced a `nodoc` method that no-ops for MySQL and PG called `mark_transaction_deferred` that I call at the start of `RealTransaction#materialize!`. This allows to compose isolation and deferral in a single transaction, when needed (as in the one test), but doesn't expose any of this externally. All transactions default to `immediate` and only fixture transactions use `deferred` via `joinable`

Reply:
Alright, I refactored that further but I think it's good to go now.